libs/server/Custom/CustomObjectFunctions.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Buffers;
+using Tsavorite.core;
+
+namespace Garnet.server
+{
+    public abstract class CustomObjectFunctions
+    {
+        /// <summary>
+        /// Whether we need an initial update, given input, if item does not already exist in store
+        /// </summary>
+        /// <param name="key">Key</param>
+        /// <param name="input">Input</param>
+        /// <param name="output">Output</param>
+        public abstract bool NeedInitialUpdate(ReadOnlyMemory<byte> key, ReadOnlySpan<byte> input, ref (IMemoryOwner<byte>, int) output);
+
+        /// <summary>
+        /// Create initial value, given key and input. Optionally generate output for command.
+        /// </summary>
+        /// <param name="key">Key</param>
+        /// <param name="input">Input</param>
+        /// <param name="value">Value</param>
+        /// <param name="output">Output</param>
+        /// <param name="rmwInfo">Advanced arguments</param>
+        /// <returns>True if done, false if we need to cancel the update</returns>
+        public abstract bool InitialUpdater(ReadOnlyMemory<byte> key, ReadOnlySpan<byte> input, IGarnetObject value, ref (IMemoryOwner<byte>, int) output, ref RMWInfo rmwInfo);
+
+        /// <summary>
+        /// Update given value in place, given key and input. Optionally generate output for command.
+        /// </summary>
+        /// <param name="key">Key</param>
+        /// <param name="input">Input</param>
+        /// <param name="value">Value</param>
+        /// <param name="output">Output</param>
+        /// <param name="rmwInfo">Advanced arguments</param>
+        /// <returns>True if done, false if we have no space to update in place</returns>
+        public abstract bool InPlaceUpdater(ReadOnlyMemory<byte> key, ReadOnlySpan<byte> input, IGarnetObject value, ref (IMemoryOwner<byte>, int) output, ref RMWInfo rmwInfo);
```

Comment:
Merge both IPU and CU into a single Updater as the copying part will be done by us in framework before calling Updater().

Reply:
As discussed, leaving both IPU and CU in this class to maintain symmetry with CustomRawStringFunctions. However, the CopyUpdater will have a default implementation of invoking the InPlaceUpdater.